### PR TITLE
Added more detail

### PR DESCRIPTION
**Pull request by zainafzal08**
Started to add more detail

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -57,95 +57,224 @@ Some of the skills/concepts this assignment aims to test (and build upon):
...
+
+> POST `/signup`
 
-200 OK for successful login
-403 for invalid credentials
-POST /signup
+```json
 {
   "displayName": "string",
   "username": "string",
```
Comment (DarkPurple141) [inline README.md]: email over displayName and username?

File: README.md [hunk 2]
```diff
@@ -57,95 +57,224 @@ Some of the skills/concepts this assignment aims to test (and build upon):
...
   "password": "string",
 }
-returns
+```
+
+> POST `/feed`
```
Comment (DarkPurple141) [inline README.md]: This fulfills the semantics of a GET not a post.
A POST is always used for the creation of a new entry.

File: README.md [hunk 3]
```diff
@@ -57,95 +57,224 @@ Some of the skills/concepts this assignment aims to test (and build upon):
...
+}
+```
+
+returns:
+  - `200` OK for successful fetch, returns a list of feed items and a position token.
+    - position token allows a client to request 10 more posts from the last post requested onwards.
+  - `403` if user not logged in or invalid position supplied
+
+> POST `/like`
+
```
Comment (DarkPurple141) [inline README.md]: PUT or POST. Also is it worth having this many API touchpoints. More common would be `/post/like` or `/post/{id}/like`

File: README.md [hunk 4]
```diff
@@ -57,95 +57,224 @@ Some of the skills/concepts this assignment aims to test (and build upon):
...
+{
+  "postId": "integer"
+}
+```
+
+returns:
+  - `200` OK for successful like
+  - `403` if user not logged in or postId supplied is invalid
+
+> POST `/follow`
```
Comment (DarkPurple141) [inline README.md]: `/user/follow` also potentially should be a PUT.

File: README.md [hunk 5]
```diff
@@ -57,95 +57,224 @@ Some of the skills/concepts this assignment aims to test (and build upon):
...
+  "username": "string"
+}
+```
+
+returns:
+  - `200` OK for successful follow
+  - `403` if user not logged in or supplied username invalid
+
+> POST `/unfollow`
+
```
Comment (DarkPurple141) [inline README.md]: DELETE `user/follow`

Comment (DarkPurple141) [reply README.md]: OR PUT. Probably PUT actually

File: README.md [hunk 6]
```diff
@@ -57,95 +57,224 @@ Some of the skills/concepts this assignment aims to test (and build upon):
...
+  "username": "string"
+}
+```
+
+returns:
+  - `200` OK for successful unfollow
+  - `403` if user not logged in or supplied username invalid or currently unfollowed
+
+
+> POST `/user/info`
```
Comment (DarkPurple141) [inline README.md]: GET `/user?id=?`

File: README.md [hunk 7]
```diff
@@ -57,95 +57,224 @@ Some of the skills/concepts this assignment aims to test (and build upon):
...
+```
+
+returns:
+  - `200` OK for successful user info fetch, returns the user info object
+  - `403` if user not logged in
+
+
+> POST `/user/update`
+
+```json
```
Comment (DarkPurple141) [inline README.md]: PUT. Maybe `/user` .. nobody else but self can edit user. So assumes logged in user, and then credentials should be forwarded, or fetched via HTTP-cookie.

File: README.md [hunk 8]
```diff
@@ -57,95 +57,224 @@ Some of the skills/concepts this assignment aims to test (and build upon):
...
+{
+  ...TBA user object
+}
+```
+
+returns:
+  - `200` OK for successful update, matches db version of user with supplied user object
+  - `403` if user not logged in or supplied update values invalid
+
+> POST `/comment`
```
Comment (DarkPurple141) [inline README.md]: POST is fine but maybe `/post/comment` as comments will never not live inside a post.

**Review by DarkPurple141 — CHANGES_REQUESTED** [review]
https://www.restapitutorial.com/ ;-)

File: README.md (hunk 2 above)
Comment (zainafzal08) [reply README.md]: Using a message body with GET is generally discouraged. Shifting to URL parameter based request.

File: README.md (hunk 1 above)
Comment (DarkPurple141) [reply README.md]: username && displayName is overkill. What's the difference? Just more data. should be email, username, password

Comment (zainafzal08) [reply README.md]: I dunno dude you put displayName in the original spec i just left it there

File: README.md (hunk 2 above)
Comment (DarkPurple141) [reply README.md]: No body necessary. get `/feed` will be resolved by HTTP-cookie. Isn't it always the current users feed?

File: README.md (hunk 1 above)
Comment (DarkPurple141) [reply README.md]: XD. blame @DarkPurple141. I'm not always right, but defs remove duplication.

File: README.md (hunk 2 above)
Comment (zainafzal08) [reply README.md]: We need to provide some sort of way to allow for students to request n items unless each request returns the entire feed

Comment (DarkPurple141) [reply README.md]: `/feed?p=2`

`/feed` provides say 10 most recent?

Comment (DarkPurple141) [reply README.md]: Then each p, just gives the next batch?

Comment (zainafzal08) [reply README.md]: Yep that's the current system.

Comment (DarkPurple141) [reply README.md]: Maybe /feed returns all. WHich we'll allow for people who can't deal with small request bundles in their logic. Then p=1,2,3,4...n is each batch of 10.

Comment (zainafzal08) [reply README.md]: hmm yeah ok i'm ok with changing /feed to be all posts
Also the current "position token" i have in the spec fulfils what you want there, i'll just change it from a string to a integer.

File: README.md [hunk 9]
```diff
@@ -17,7 +17,7 @@ In addition we've provided a basic project scaffold for you to build from.
 You can use everything we've given you, although there's no requirement to use anything.
 ```bash
 # scaffold
-data 
+data
   - feed.json  # A sample feed data object
```
Comment (DarkPurple141) [inline README.md:21]: Is this different at all in the PR?

Comment (DarkPurple141) [reply README.md:21]: In terms of the filenames, and what's in them?

File: README.md [hunk 10]
```diff
@@ -57,95 +57,388 @@ Some of the skills/concepts this assignment aims to test (and build upon):
...
+**Response**
+`200` On successful signup, sets a cookie with a session token
+```json
+{
+  "error": false,
+  "errorMsg": null
+}
+```
+
+`403` for existent username
```
Comment (DarkPurple141) [inline README.md]: *existant

File: README.md [hunk 11]
```diff
@@ -57,95 +57,388 @@ Some of the skills/concepts this assignment aims to test (and build upon):
...
+`403` for Non Existent post or user not logged in
+```json
+{
+  "error": true,
+  "errorMsg": "You Are Not Logged In | Invalid Post Id"
+}  
+```
+
+
+> PUT `/post/comment`
```
Comment (DarkPurple141) [inline README.md]: Should be POST. "creating an entry". Maybe to standardise any id should be url param?

File: README.md [hunk 12]
```diff
@@ -57,95 +57,388 @@ Some of the skills/concepts this assignment aims to test (and build upon):
...
+`403` for user is not logged in or invalid post id
+```json
+{
+  "error": true,
+  "errorMsg": "You Are Not Logged In | Invalid Post Id"
+}  
+```
+
+> POST `/post/new`
+
```
Comment (DarkPurple141) [inline README.md]: POST `/post` -- new is redundant.
DELETE `/post?id=xxx` -- should also be a thing.

Then you've got get, create, delete, which is basic CRUD.

File: README.md [hunk 13]
```diff
@@ -57,95 +57,388 @@ Some of the skills/concepts this assignment aims to test (and build upon):
...
+{
+  "error": false,
+  "errorMsg": ""
+}
+```
+
+`403` for user is not logged in or image/description invalid
+```json
+{
+  "error": true,
```
Comment (DarkPurple141) [inline README.md]: So just standardise the error responses to simplify.

A lot are doubling up. Just keep it honouring the 40x style.

```
400 bad request (forgot a field, malformed request)
401 not authorized (not logged in)
403 forbidden (logged in, but not allowed to ser resource)
```

Comment (DarkPurple141) [reply README.md]: Also errorMsg => message

File: README.md [hunk 14]
```diff
@@ -57,95 +57,388 @@ Some of the skills/concepts this assignment aims to test (and build upon):
...
-403 for user already exists
+`403` for invalid username or user is not logged in
+```json
+{
+  "error": true,
+  "errorMsg": "You Are Not Logged In | Invalid Username"
+}  
+```
+
+
```
Comment (DarkPurple141) [inline README.md]: Just use same API route for both. Also I reckon again as per above, use standardised id for param.
Let the verbs do the backend work for you.
```
# both here are idempotent. You can't create entry 10 twice, nor can u delete it twice. Same effect.
POST     /user/follow?id=10 # creates a new follow entry
DELETE /user/follow?id=10  # deletes a follow entry
```

File: README.md [hunk 15]
```diff
@@ -57,95 +57,388 @@ Some of the skills/concepts this assignment aims to test (and build upon):
...
 
+`403` for invalid username or user is not logged in
+```json
+{
+  "error": true,
+  "errorMsg": "You Are Not Logged In | Invalid Username"
+}  
+```
+
+> GET `/user/info`
```
Comment (DarkPurple141) [inline README.md]: remove the word info. GET is enough plus id of user.
`/user?id=10`

File: README.md [hunk 16]
```diff
@@ -57,95 +57,388 @@ Some of the skills/concepts this assignment aims to test (and build upon):
...
+
+`403` for user is not logged in
+```json
+{
+  "error": true,
+  "errorMsg": "You Are Not Logged In"
+}  
+```
+
+> PUT `/user/update`
```
Comment (DarkPurple141) [inline README.md]: PUT implies update, remove /update

Just: 
`PUT /user` is enough in this instance as only current user can update current user.

File: README.md [hunk 17]
```diff
@@ -57,95 +57,388 @@ Some of the skills/concepts this assignment aims to test (and build upon):
...
+Users can write comments on "posts" via `post/comment`
 
-### Update Profile
-Users can update their personal profile. E.g:
+**Live Update**
+If a user likes a post or comments on a post, the posts likes and comments should
+update without requiring a page reload/refresh.
+
+**Update Profile**
+Users can update their personal profile via `/user/update` E.g:
```
Comment (DarkPurple141) [inline README.md]: As above.

File: README.md [hunk 18]
```diff
@@ -57,95 +57,388 @@ Some of the skills/concepts this assignment aims to test (and build upon):
...
-### Pagination (Level 2)
-Users can page between sets of results in the feed
+**Pagination**
+Users can page between sets of results in the feed using the position token from `user/feed`
 
-### Update Password (Level 2)
-Users can change their password
+**Profile**
+Users can see their own profile information such as username, number of posts, number of likes, profile pic.
+get this information from `/user/info`
```
Comment (DarkPurple141) [inline README.md]: As above.

File: README.md [hunk 19]
```diff
@@ -57,95 +57,388 @@ Some of the skills/concepts this assignment aims to test (and build upon):
...
+Allow an option for a user to see all the comments on a post.
+same as above.
+
+**Like user generated content**
+A logged in user can like a post on their feed and trigger a api request `/post/like`
+For now it's ok if the like doesn't show up until the page is refreshed.
 
-### "Post" new content (Level 2)
-Users can upload and post new content on their profile and have it appear in other users feeds.
+**"Post" new content**
```
Comment (DarkPurple141) [inline README.md]: As above.

**Review by DarkPurple141 — CHANGES_REQUESTED** [review]
Sorry more changes :). Close.

File: README.md (hunk 9 above)
Comment (zainafzal08) [reply README.md:21]: Yeah there's a small change, updating now

File: README.md (hunk 12 above)
Comment (zainafzal08) [reply README.md]: Deleting a post isn't in the current spec outline, do you want to add that?

File: README.md (hunk 14 above)
Comment (zainafzal08) [reply README.md]: I honestly think "deleting" a follow entry is super unintuitive and unnecessary here. You are updating the list of people a user follows so having a follow and unfollow endpoint makes sense.

File: README.md (hunk 13 above)
Comment (zainafzal08) [reply README.md]: I liked the idea of having a set variable json error response, It means students get meaningful error messages when debugging rather then having to reference the docs.

File: README.md (hunk 15 above)
Comment (zainafzal08) [reply README.md]: yep sounds hella

File: README.md (hunk 13 above)
Comment (DarkPurple141) [reply README.md]: You can still provide that but tbh you're creating more work, and more routes to test and debug.

File: README.md (hunk 14 above)
Comment (DarkPurple141) [reply README.md]: Alright, PUT then

File: README.md (hunk 12 above)
Comment (DarkPurple141) [reply README.md]: Don't mind, can leave if you want.

File: README.md (hunk 13 above)
Comment (zainafzal08) [reply README.md]: Fair point, I guess let's stick with the standard 40x style for now then and if we have time we can give more meaningful error objects later on

**Review by DarkPurple141 — APPROVED** [review]

**Pull request merged**